src/Stratis.Bitcoin.Features.SmartContracts/Embedded/Multisig.cs

```diff
@@ -0,0 +1,152 @@
+﻿using System.Linq;
+using NBitcoin;
+using Stratis.SmartContracts;
+using Stratis.SmartContracts.CLR;
+
+[EmbeddedContract(EmbeddedContractType.Multisig)]
+public class MultiSig : SmartContract
+{
+    const string primaryGroup = "main";
+    private readonly uint version;
+    private readonly Authentication authentication;
+
+    public MultiSig(ISmartContractState state, IPersistenceStrategy persistenceStrategy, Network network) : base(state)
+    {
+        uint version = state.Message.ContractAddress.ToUint160().GetEmbeddedVersion();
+
+        Assert(version == 1, "Only a version of 1 is supported.");
+
+        this.version = version;
+        this.authentication = new Authentication(GetState(state, persistenceStrategy, EmbeddedContractAddress.Create(typeof(Authentication), 1)), network);
+
+        // Exit if already initialized.
+        if (this.Initialized)
+            return;
+
+        foreach (IFederation federation in network.Federations.GetFederations())
+        {
+            (PubKey[] pubKeys, int signaturesRequired) = federation.GetFederationDetails();
+            AddFederation(federation.Id.ToHex(network), (uint)signaturesRequired, pubKeys.Select(pk => pk.ToHex()).ToArray());
+        }
+
+        this.Initialized = true;
+    }
+
+    private ISmartContractState GetState(ISmartContractState state, IPersistenceStrategy persistenceStrategy, uint160 address)
+    {
+        return new SmartContractState(state.Block, state.Message, new PersistentState(persistenceStrategy, state.Serializer, address),
+            state.Serializer, state.ContractLogger, state.InternalTransactionExecutor, state.InternalHashHelper, state.GetBalance);
+    }
+
+    public bool Initialized
+    {
+        get => this.State.GetBool("Initialized");
+        set => this.State.SetBool("Initialized", value);
+    }
+
+    private void VerifySignatures(string group, byte[] signatures, string authorizationChallenge)
+    {
+        this.authentication.VerifySignatures(group, signatures, authorizationChallenge);
+    }
+
+    private void AddFederation(string federationId, uint quorum, string[] pubKeys)
+    {
+        SetFederationMembers(federationId, pubKeys);
+        SetFederationQuorum(federationId, quorum);
+    }
+
+    public string[] GetFederationMembers(string federationId)
+    {
+        Assert(!string.IsNullOrEmpty(federationId));
+        return this.State.GetArray<string>($"Members:{federationId}");
+    }
+
+    private void SetFederationMembers(string federationId, string[] values)
+    {
+        this.State.SetArray($"Members:{federationId}", values);
```

Comment:
Use consistent naming for keys plural or singular. My recommendation is keep it singular.